src/librustc/hir/mod.rs

```diff
@@ -145,7 +145,27 @@ pub struct Lifetime {
     /// HIR lowering inserts these placeholders in type paths that
     /// refer to type definitions needing lifetime parameters,
     /// `&T` and `&mut T`, and trait objects without `... + 'a`.
-    pub name: Name,
+    pub name: LifetimeName,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Copy)]
+pub enum LifetimeName {
+    Elided,
+    Underscore,
+    Static,
+    Name(Name),
+}
+
+impl LifetimeName {
+    pub fn name(&self) -> Name {
+        use self::LifetimeName::*;
+        match *self {
+            Elided => keywords::Invalid.name(),
+            Underscore => Symbol::intern("'_"),
+            Static => Symbol::intern("'static"),
+            Name(name) => name,
```

Comment:
Nooooo.
This `LifetimeName` over `Name` is pure over-engineering, N extra lines of boilerplate just to avoid one string comparison.

Reply:
Haha. I don't have any strong opinion here. Take it up with @nikomatsakis :smile:
I left it as it's own commit so it's easy to pull out if that's the decision.

Reply:
I'd better turn `'_` into a weak keyword.

Reply:
I disagree. Sentinel values mean you wind up with if-else-if chains like `if foo.is_elided() { ... }` which are easily forgotten. Having an enum means we can use exhaustive matching, so that when we add variants in the future, the compiler helps us catch cases that are missing etc.

Plus, there isn't a lot of boilerplate here.

Reply:
@nikomatsakis 
Ok, not a big deal.
(I'll revert this when you are on vacation, ha-ha.)